File: app/src/main/java/sfllhkhan95/doodle/projects/utils/ThumbnailInflater.java
package sfllhkhan95.doodle.projects.utils;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import sfllhkhan95.doodle.R;
import sfllhkhan95.doodle.core.MainActivity;
import sfllhkhan95.doodle.projects.models.Thumbnail;

@UiThread
public class ThumbnailInflater implements Runnable, AdapterView.OnItemClickListener {

    private final Activity activity;
    private String[] savedProjects;

    public ThumbnailInflater(Activity activity) {
        this.activity = activity;
    }

    public void setSavedProjects(String[] savedProjects) {
        this.savedProjects = savedProjects;
    }

    @NonNull
    private List<Thumbnail> getThumbnails() {
        List<Thumbnail> thumbnails = new ArrayList<>();
        if (savedProjects != null) {
            for (String projectName : savedProjects) {
                Bitmap thumbnailBitmap;
                if ((thumbnailBitmap = DoodleDatabase.loadDoodle(projectName, 200, 200)) != null) {
                    Thumbnail thumbnail = new Thumbnail(this, thumbnailBitmap, projectName);
                    thumbnails.add(thumbnail);
                }
            }
        }
        return thumbnails;
    }

    private void inflateGrid() {
        List<Thumbnail> thumbnails = getThumbnails();
        ThumbnailAdapter adapter = new ThumbnailAdapter(
                activity,
                R.layout.template_thumbnail,
                thumbnails);

        GridView projectGrid = activity.findViewById(R.id.savedProjectsGrid);
        projectGrid.setAdapter(adapter);

        projectGrid.setOnItemClickListener(this);
    }

    private void inflateList() {
        List<Thumbnail> thumbnails = getThumbnails();
        ThumbnailAdapter adapter = new ThumbnailAdapter(
                activity,
                R.layout.template_thumbnail_list,
                thumbnails);

        ListView projectList = activity.findViewById(R.id.savedProjectsList);
        projectList.setAdapter(adapter);
        projectList.setOnItemClickListener(this);
    }

    @Override
    public void run() {
        ((TextView) activity.findViewById(R.id.postCount)).setText(String.valueOf(getThumbnails().size()));
        inflateGrid();
        inflateList();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        Intent intent = new Intent(activity, MainActivity.class);
        Thumbnail item = (Thumbnail) parent.getItemAtPosition(position);
        intent.putExtra("DOODLE", item.getName());

        if (!DoodleDatabase.contains(item.getName())) return;

        activity.startActivity(intent);
    }

}

File: app/src/main/java/sfllhkhan95/doodle/projects/utils/ThumbnailInflater.kt
package sfllhkhan95.doodle.projects.utils

import android.app.Activity
import android.content.Intent
import android.support.annotation.UiThread
import android.util.DisplayMetrics
import android.view.View
import android.widget.AdapterView
import android.widget.GridView
import sfllhkhan95.doodle.R
import sfllhkhan95.doodle.core.MainActivity
import sfllhkhan95.doodle.projects.models.Thumbnail
import java.util.*

@UiThread
class ThumbnailInflater(private val activity: Activity) : Runnable, AdapterView.OnItemClickListener {
    private var savedProjects: Array<String>? = null

    private val thumbnails: List<Thumbnail>
        get() {
            val thumbnails = ArrayList<Thumbnail>()
            if (savedProjects != null) {
                for (projectName in savedProjects!!) {
                    val thumbnailBitmap = DoodleDatabase.loadDoodle(projectName, 200, 200)
                    if (thumbnailBitmap != null) {
                        val thumbnail = Thumbnail(thumbnailBitmap, projectName)
                        thumbnails.add(thumbnail)
                    }
                }
            }
            return thumbnails
        }

    private// Obtain device display metrics (used to setup project resolution)
    val thumbnailsFull: List<Thumbnail>
        get() {
            val thumbnails = ArrayList<Thumbnail>()
            if (savedProjects != null) {
                for (projectName in savedProjects!!) {
                    val metrics = DisplayMetrics()
                    activity.windowManager.defaultDisplay.getMetrics(metrics)

                    val thumbnailBitmap = DoodleDatabase.loadDoodle(projectName, metrics.widthPixels / 2, metrics.heightPixels / 2)
                    if (thumbnailBitmap != null) {
                        val thumbnail = Thumbnail(thumbnailBitmap, projectName)
                        thumbnails.add(thumbnail)
                    }
                }
            }
            return thumbnails
        }

    fun setSavedProjects(savedProjects: Array<String>?) {
        this.savedProjects = savedProjects
    }

    private fun inflateGrid() {
        val thumbnails = thumbnails
        val adapter = ThumbnailAdapter(
                activity,
                R.layout.template_thumbnail,
                thumbnails)

        val projectGrid = activity.findViewById<GridView>(R.id.savedProjectsGrid)
        projectGrid.adapter = adapter

        projectGrid.onItemClickListener = this
    }

    override fun run() {
        inflateGrid()
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val intent = Intent(activity, MainActivity::class.java)
        val item = parent.getItemAtPosition(position) as Thumbnail
        intent.putExtra("DOODLE", item.name)
        intent.putExtra("READ_ONLY", true)

        if (!DoodleDatabase.contains(item.name!!)) return

        activity.startActivity(intent)
    }

    fun share(name: String) {
        val intent = Intent(activity, MainActivity::class.java)
        intent.putExtra("DOODLE", name)
        intent.putExtra("SHARE", true)
        intent.putExtra("READ_ONLY", true)

        if (!DoodleDatabase.contains(name)) return

        activity.startActivity(intent)
    }
}